### PR TITLE
*: support show stats_buckets

### PR DESCRIPTION
**Pull request by lamxTyler**
The syntax is `show stats_buckets` or `show stats_buckets where ...`.
For example:
```
MySQL [test]> show stats_buckets;
+---------+------------+-------------+----------+-----------+-------+---------+-------------+-------------+
| Db_name | Table_name | Column_name | Is_index | Bucket_id | Count | Repeats | Lower_Bound | Upper_Bound |
+---------+------------+-------------+----------+-----------+-------+---------+-------------+-------------+
| test    | t          | a           |        0 |         0 |     1 |       1 | 1           | 1           |
| test    | t          | b           |        0 |         0 |     1 |       1 | 1           | 1           |
| test    | t          | idx         |        1 |         0 |     1 |       1 | (1, 1)      | (1, 1)      |
+---------+------------+-------------+----------+-----------+-------+---------+-------------+-------------+
```
PTAL @hanfei1991 @zimulala  @winoros

### CODE REVIEW

File: executor/show.go [hunk 1]
```diff
@@ -694,3 +697,91 @@ func (e *ShowExec) versionToTime(version uint64) types.Time {
 	t := time.Unix(0, oracle.ExtractPhysical(version)*int64(time.Millisecond))
 	return types.Time{Time: types.FromGoTime(t), Type: mysql.TypeDatetime}
 }
+
+func (e *ShowExec) fetchShowStatsBuckets() error {
+	do := sessionctx.GetDomain(e.ctx)
+	h := do.StatsHandle()
+	dbs := do.InfoSchema().AllSchemas()
```
Comment (hanfei1991) [inline executor/show.go]: Can we implement a `StatsCacheToRows` function for handle instead of fetch them through infoschema?

Comment (hanfei1991) [inline executor/show.go]: We can create a file named `statistics/show.go`

Comment (lamxTyler) [inline executor/show.go]: I didn't find a way to get the `db_name` from a `tableId`, so I implement it this way.

Comment (hanfei1991) [inline executor/show.go]: We can store the dbName when building `statistics.Table`.

Comment (lamxTyler) [inline executor/show.go]: In fact, when we build the `statistics.Table`, all we have is only the `tableID`.

Comment (hanfei1991) [inline executor/show.go]: OK, let's remain it. Maybe we have some better way in the future.

File: executor/show_stats_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+func (s *testSuite) TestShowStatsBuckets(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int)")
+	tk.MustExec("create index idx on t(a,b)")
+	tk.MustExec("insert into t values (1,1)")
+	tk.MustExec("analyze table t")
+	result := tk.MustQuery("show stats_buckets").Sort()
+	c.Assert(len(result.Rows()), Equals, 3)
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: Why not use result.Check(testkit.Rows(...))

File: executor/show_stats_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+func (s *testSuite) TestShowStatsBuckets(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int)")
+	tk.MustExec("create index idx on t(a,b)")
+	tk.MustExec("insert into t values (1,1)")
+	tk.MustExec("analyze table t")
+	result := tk.MustQuery("show stats_buckets").Sort()
+	result.Check(testkit.Rows("test t idx 1 0 1 1 (1, 1) (1, 1)] [test t a 0 0 1 1 1 1] [test t b 0 0 1 1 1 1"))
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: Are `test t idx 1 0 1 1 (1, 1) (1, 1)` , `[test t a 0 0 1 1 1 1]` and `test t b 0 0 1 1 1 1` three different lines ?
Should it check as `testkit.Rows("test t idx 1 0 1 1 (1, 1) (1, 1)", "test t a 0 0 1 1 1 1", "test t b 0 0 1 1 1 1")`

File: executor/show_stats_test.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+func (s *testSuite) TestShowStatsMeta(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t, t1")
+	tk.MustExec("create table t (a int, b int)")
+	tk.MustExec("create table t1 (a int, b int)")
+	tk.MustExec("analyze table t, t1")
+	result := tk.MustQuery("show stats_meta")
+	c.Assert(len(result.Rows()), Equals, 2)
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: Also need to change this ....

File: executor/show_stats_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	c.Assert(result.Rows()[0][1], Equals, "t")
+	c.Assert(result.Rows()[1][1], Equals, "t1")
+	result = tk.MustQuery("show stats_meta where table_name = 't'")
+	c.Assert(len(result.Rows()), Equals, 1)
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: And this

File: executor/show_stats_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (s *testSuite) TestShowStatsHistograms(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int)")
+	tk.MustExec("analyze table t")
+	result := tk.MustQuery("show stats_histograms").Sort()
+	c.Assert(len(result.Rows()), Equals, 2)
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: And this

File: executor/show_stats_test.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	c.Assert(result.Rows()[0][2], Equals, "a")
+	c.Assert(result.Rows()[1][2], Equals, "b")
+	result = tk.MustQuery("show stats_histograms where column_name = 'a'")
+	c.Assert(len(result.Rows()), Equals, 1)
```
Comment (hanfei1991) [inline executor/show_stats_test.go]: And this

File: executor/show_stats_test.go (hunk 4 above)
Comment (lamxTyler) [inline executor/show_stats_test.go]: This contains timestamp, so it will be different every time.

Comment (hanfei1991) [inline executor/show_stats_test.go]: Ok

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@winoros PTAL

File: executor/show_stats.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+func (e *ShowExec) valueToString(value types.Datum, size int) (string, error) {
+	if size == 0 {
+		return value.ToString()
+	}
+	decodedVals, err := codec.Decode(value.GetBytes(), size)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	var strs []string
+	for _, val := range decodedVals {
```
Comment (winoros) [inline executor/show_stats.go]: I think this can be extracted as function `DatumsToString` in package `types.datum` or there already is one?

**Pull request merged**